text/0000-gen-fn.md

```diff
@@ -0,0 +1,429 @@
+
+```rust
+#![feature(generators)]
+#![feature(iter_from_generator)]
+
+fn main() {
+    let mut it = std::iter::from_generator(|| {
+        yield 1
+    });
+
+    assert_eq!(it.next(), Some(1));
+    assert_eq!(it.next(), None);
+    it.next(); // panics
+}
+```
+
+## Panicking
+
+What happens when `Iterator::next` is called again on a `#[rustc_gen]` block that panicked? Do we need to poison the iterator?
+
+## Contextual keyword
+
+Popular crates (like `rand`) have methods called [`gen`][Rng::gen]. If we forbid those, we are forcing those crates to make a major version bump when they update their edition, and we are requiring any users of those crates to use `r#gen` instead of `gen` when calling that method.
+
+We could choose to use a contextual keyword and only forbid `gen` in
+
+* bindings,
+* field names (due to destructuring bindings),
+* enum variants,
+* and type names
+
+This should avoid any parsing issues around `gen` followed by `{` in expressions.
+
+[Rng::gen]: https://docs.rs/rand/latest/rand/trait.Rng.html#method.gen
+
+## `Iterator::size_hint`
+
+Should we try to compute a conservative `size_hint`? This will reveal information from the body of a generator,
+but at least for simple cases users will likely expect `size_hint` to not just be the default.
+
+## Implement other `Iterator` traits.
+
+Is there a possibility for implementing traits like `DoubleEndedIterator`, `ExactSizeIterator` at all?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## `yield from` (forwarding operation)
+
+Python has the ability to `yield from` an iterator.
+Effectively this is syntax sugar for looping over all elements of the iterator and yielding them individually.
+There are infinite options to choose from if we want such a feature, so I'm listing general ideas:
+
+### Do nothing, just use loops
+
+```rust
+for x in iter {
+    yield x
+}
+```
+
+### Language support
+
+we could do something like postfix `yield` or an entirely new keyword, or...
+
+```rust
+iter.yield
+```
+
+### stdlib macro
+
+We could add a macro to the standard library and prelude.
+The macro would expand to a `for` loop + `yield`.
+
+```rust
+yield_all!(iter)
+```
+
+## Complete `Generator` support
+
+We already have a `Generator` trait on nightly that is more powerful than the `Iterator`
+API could possibly be.
+
+1. it uses `Pin<&mut Self>`, allowing self-references in the generator across yield points
+2. it has arguments (`yield` returns the arguments passed to it in the subsequent invocations)
+
+Similar to the ideas around `async` closures,
+I think we could argue for `Generators` to be `#[rustc_gen]` closures while `#[rustc_gen]` blocks are a simpler concept that has no arguments and only captures variables.
+
+Either way, support for full `Generator`s should be discussed and implemented separately,
+as there are many more open questions around them beyond a simpler way to write `Iterator`s.
```

Comment:
We should explicitly mention the possibility of writing

```
gen {
    let x = yield 0;
    println!("{x:}");
}
```
which is supported by `Generator::resume(val)`, but isn't by `Iterator` (today? we could add a new method `Iterator::resume` that by default calls `Iterator::next`).

Reply:
That behaviour needs a more closure like syntax (like nightly generators) anyway, as you also need the argument before the first yield. Even more syntax will make it easier to distinguish between generators and coroutines or anything in between

Reply:
I'll add a note to the full generators section